zaza/utilities/openstack_provider.py

```diff
@@ -0,0 +1,305 @@
+# Copyright 2021 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Functions to work with the openstack provider."""
+
+import logging
+import os
+import tenacity
+
+from keystoneauth1 import session
+from keystoneauth1.identity import (
+    v3,
+    v2,
+)
+from novaclient import client as novaclient_client
+
+
+class MissingOSAthenticationException(Exception):
+    """Exception when some data needed to authenticate is missing."""
+
+    pass
+
+
+def get_undercloud_keystone_session(verify=None):
+    """Return Under cloud keystone session.
+
+    :param verify: Control TLS certificate verification behaviour
+    :type verify: any
+    :returns keystone_session: keystoneauth1.session.Session object
+    :rtype: keystoneauth1.session.Session
+    """
+    return get_keystone_session(get_undercloud_auth(),
+                                verify=verify)
+
+
+def get_keystone_session(openrc_creds, scope='PROJECT', verify=None):
+    """Return keystone session.
+
+    :param openrc_creds: OpenStack RC credentials
+    :type openrc_creds: dict
+    :param verify: Control TLS certificate verification behaviour
+    :type verify: any (True  - use system certs,
+                       False - do not verify,
+                       None  - defer to requests library to find certs,
+                       str   - path to a CA cert bundle)
+    :param scope: Authentication scope: PROJECT or DOMAIN
+    :type scope: string
+    :returns: Keystone session object
+    :rtype: keystoneauth1.session.Session object
+    """
+    keystone_creds = get_ks_creds(openrc_creds, scope=scope)
+    if not verify and openrc_creds.get('OS_CACERT'):
```

Comment:
I think we should enter here only if `verify` is `None` but not if it's `False`, so `s/if not verify/if verify is None/`?

Or is it code that you copy/pasted from somewhere else?

Reply:
This code is c/p from the zaza-openstack-tests code; I don't really want to mess with it as it seems to work.